contrib/epee/include/net/net_helper.h

```diff
@@ -302,14 +336,17 @@ namespace net_utils
 				*/
 				boost::system::error_code ec;
 
-				size_t writen = m_socket.write_some(boost::asio::buffer(data, sz), ec);
+				size_t writen;
+				if(m_ssl)
+					writen = m_ssl_socket.write_some(boost::asio::buffer(data, sz), ec);
+				else
+					writen = m_socket.write_some(boost::asio::buffer(data, sz), ec);
```

Comment:
There is no common base class. Crap. The only other way to get around this would be to something like:
```c++
template<typename Socket>
void do_run(Socket& socket) {
    ... all code here ...
}

void run() {
    if (m_ssl) do_run(m_ssl_socket) else do_run(m_ssl_socket.next_layer());
}
```
but then will force the compiler to create two copies of the function in the assembly output, and they will be nearly identical except for the few a calls into ASIO. Another option (which I also dislike) would be to create a custom base class with two derived types to inject polymorphism into the function. If done properly there would be no memory allocation over head (only one stack pointer for the vtable would be necessary), but there is lots of boilerplate.

I think the current implementation is probably the best of all the options.